File: System.Automata/Transition.cs
namespace System.Automata {
	/// <summary>
	/// A state transition
	/// </summary>
	public struct Transition {
		/// <summary>
		/// The current state.
		/// </summary>
		public State P { get; }
		
		/// <summary>
		/// The next state.
		/// </summary>
		public State Q { get; }
		
		/// <summary>
		/// The symbol to transition on
		/// </summary>
		public char? A { get; }
		

		/// <param name="p">The current state</param>
		/// <param name="c">The current input symbol</param>
		/// <param name="q">The next state</param>
		/// <remarks>For a epsilon-transition, use Alphabet.<see cref="Alphabet.EmptyString" /></remarks>
		public Transition(State p, char? c, State q) {
			P = p;
			A = c;
			Q = q;
		}

		public override string ToString() {
			return $"δ({P}, {A.ToString().Replace('\0'.ToString(), "ε")}) = {Q}";
		}
	}
}

File: src/Transition.cpp
#include "Transition.h"

namespace Automata {
	State* Transition::getCurrent() {
		return _p;
	}

	State* Transition::getNext() {
		return _q;
	}

	char Transition::getSymbol() {
		return _a;
	}

	Transition::Transition(State *p, State *q, char s) {
		_p = p;
		_q = q;
		_a = s;
	}

	Transition::operator string(){
		stringstream ret;
		ret << "δ(" << _p << ", ";
		if(_a == EMPTY_STRING) ret << "ε";
		else ret << _a;
		ret << ") = " << _q;
		return ret.str();
	}

	bool Transition::operator==(Transition* a) {
		if(a == this) return true;
		return a->getCurrent() == this->getCurrent()
		       && a->getNext() == this->getNext()
		       && a->getSymbol() == this->getSymbol();
	}
}
